micro-rdk/src/common/webrtc/signaling_server.rs

```diff
@@ -0,0 +1,176 @@
+use super::api::{WebRtcSdp, WebRtcSignalingChannel};
+use crate::{
+    common::{
+        exec::Executor,
+        grpc::{GrpcError, ServerError},
+        webrtc::api::{SdpOffer, WebRtcError},
+    },
+    proto::rpc::webrtc::v1::{
+        CallRequest, CallResponse, CallUpdateRequest, CallUpdateResponse,
+        OptionalWebRtcConfigRequest, OptionalWebRtcConfigResponse,
+    },
+};
+use async_channel::{RecvError, Sender};
+use base64::{engine::general_purpose, Engine};
+use either::Either;
+use std::{collections::HashMap, io::Cursor, sync::Mutex};
+use uuid::Uuid;
+
+pub(crate) struct LocalSignaling {
+    pub(crate) tx: async_channel::Sender<Result<CallResponse, ServerError>>,
+    pub(crate) rx: async_channel::Receiver<CallUpdateRequest>,
+}
+
+pub(crate) struct SignalingServer {
+    pub executor: Executor,
+    sender: Sender<Box<WebRtcSignalingChannel>>,
+    pending: Mutex<HashMap<String, async_channel::Sender<CallUpdateRequest>>>,
+}
+
+impl SignalingServer {
+    pub fn new(executor: Executor, sender: Sender<Box<WebRtcSignalingChannel>>) -> Self {
+        Self {
+            executor,
+            sender,
+            pending: Mutex::new(HashMap::new()),
+        }
+    }
+
+    pub fn optional_webrtc_config(
+        &self,
+        _request: OptionalWebRtcConfigRequest,
+    ) -> Result<OptionalWebRtcConfigResponse, ServerError> {
+        Ok(OptionalWebRtcConfigResponse { config: None })
+    }
+
+    pub async fn call(
+        &self,
+        request: CallRequest,
+        responses: Sender<Result<CallResponse, ServerError>>,
+    ) -> Result<(), ServerError> {
+        if request.disable_trickle {
+            return Err(ServerError::new(GrpcError::RpcUnimplemented, None));
+        }
+
+        // TODO: Can we re-use this engine?
+        let engine: Box<general_purpose::GeneralPurpose> = general_purpose::STANDARD.into();
```

Comment:
Building the SDP from the CallRequest is very similar to how we build it from the `AnswerRequest` in `wait_for_sdp`, and should be DRY'd.

Reply:
we can add a ticket for making the engine global

Reply:
Ticket added, and TODO'd: https://viam.atlassian.net/browse/RSDK-9247